[Feature] LLM collector

Lucaskabela opened the pull request:
## Description

Add LLM Collector, which is a subset of the SyncDatacollector aimed at trimming to core funcionality in an easy to consume API for LLM devs

## Motivation and Context

#2872 

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [X] New feature (non-breaking change which adds core functionality)


## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [X] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.

## Testing
```
python test/test_collector.py TestLLMCollector
```

Lucaskabela commented on torchrl/collectors/llm_collector.py:
```diff
@@ -0,0 +1,134 @@
+        while collected_frames < self.frames_per_batch:
+            policy_input = data
+            print('policy_input type', type(policy_input))
+            env_input = self.policy(policy_input)
```
seems this data causes some exception in serialization format when passed to the LLM - wondering if we need some additional parameters to the constructor for the policy? (cc @vmoens)

vmoens replied:
I'm not seeing any error on cpu (at least for 20 iterations )

Lucaskabela replied:
Interesting; assuming you are running the testplan as is @vmoens? I get the crash listed on the very first iteration

What version of vllm is in your environment out of curiosity?

Lucaskabela commented:
Originally, this code ran into the error:
```
Traceback (most recent call last):
  File "/home/lucaskabela/.conda/envs/cabernet/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/home/lucaskabela/.conda/envs/cabernet/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/home/lucaskabela/.conda/envs/cabernet/lib/python3.10/site-packages/vllm/v1/engine/core.py", line 429, in process_input_socket
    request = decoder.decode(data_frame.buffer)
  File "/home/lucaskabela/.conda/envs/cabernet/lib/python3.10/site-packages/vllm/v1/serial_utils.py", line 34, in decode
    return self.decoder.decode(obj)
msgspec.ValidationError: Expected `int | null`, got `bool` - at `$[6].logprobs`
```

This seems to have come from a versioning issue; downgrading torch and vllm with
```
pip install --force torch==2.5.1 vllm==0.7.3
```
Then
```
python setup.py clean && python setup.py develop
```
Did the trick and fixed this error :) Previously, I had `torch==2.6.0` and `vllm==0.8.2`, so note these versions may have cause this error

vmoens commented:
Amazing!
I did a bit of refactoring.
Things that are missing:
- Tests in test/test_collectors.py. These should include the vLLMWrapper and the TransformersWrapper
- docstrings + register in the `docs` directory
- ultimately remove the `if __name__ == "__main__":` from the file

Lucaskabela commented:
> Amazing! I did a bit of refactoring. Things that are missing:
> 
> * Tests in test/test_collectors.py. These should include the vLLMWrapper and the TransformersWrapper
> * docstrings + register in the `docs` directory
> * ultimately remove the `if __name__ == "__main__":` from the file
> * use `trust_policy=True` as not doing so can trigger checks that we want to avoid


Thanks for the comments here; I think I incorporated the majority of this feedback and updated the test on my end; the only thing I am not sure about is how to register in the `docs` directory - could you provide more details on this?

vmoens commented on test/test_collector.py:
```diff
@@ -3544,6 +3559,94 @@ def test_weight_update(self):
+        )
+
+        for data in collector:
+            assert data is not None
```
maybe let's check that the fields make sense?
ie that we have all the set of tokens / text etc that we want?

vmoens replied:
Another thing to check is that the data has the right number of elements and that the total number of steps matches what you set in the constructor. 
We also want to test that we can populate a replay buffer:
```
rb = ReplayBuffer(storage=LazyStackStorage(size))
```
and that the samples "make sense" (eg, they have the collector metadata under the "collector" key)

vmoens commented on test/test_collector.py:
```diff
@@ -3544,6 +3559,94 @@ def test_weight_update(self):
+        )
+
+        for data in collector:
+            assert data is not None
```
ditto

vmoens commented on torchrl/collectors/llm_collector.py:
```diff
@@ -0,0 +1,78 @@
+
+
+class LLMCollector(SyncDataCollector):
+    """A simplified version of SyncDataCollector for LLM inference."""
```
missing kwarg description + running example (feel free to pick the one in the tests)

vmoens commented on torchrl/collectors/llm_collector.py:
```diff
@@ -0,0 +1,78 @@
+        env,
+        policy_factory,
+        *,
+        steps_per_batch: int = -1,
```
```suggestion
        steps_per_batch: int,
```
Right now this is a value that we require. Long-term, when we'll go full async, we'll allow people to not specify anything because the output data will be written the buffer as soon as the trajectory is completed.

vmoens reviewed (approved):
LGTM thanks!
